// ya-ts-client/dist/ya-activity/src/api/requestor-control-api.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Yagna Activity API
 *  The Activity API can be perceived as controls which a Requestor-side application has to steer the execution of an Activity as specified in an Agreement which has been negotiated via the Market API/Protocol. This defines possible interactions between the Requestor application (via Activity API) and the generic components running on the Provider node, which host the Provider-side application code. The possible interactions imply a logical “execution environment” component, which is the host/container for the “payload” code. The “execution environment” is specified as an ExeUnit, with a generic interface via which a Provider node’s Activity Controller can operate the hosted code. It conforms with capability level 1 of the [Activity API specification] (https://docs.google.com/document/d/1BXaN32ediXdBHljEApmznSfbuudTU8TmvOmHKl0gmQM).
 *
 * The version of the OpenAPI document: 1.6.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestorControlApi = exports.RequestorControlApiFactory = exports.RequestorControlApiFp = exports.RequestorControlApiAxiosParamCreator = void 0;
const axios_1 = require("axios");
// Some imports not used depending on template conditions
// @ts-ignore
const base_1 = require("../../base");
/**
 * RequestorControlApi - axios parameter creator
 * @export
 */
exports.RequestorControlApiAxiosParamCreator = function (configuration) {
    return {
        /**
         *
         * @summary Sends encrypted command to secure exe-unit.
         * @param {string} activityId
         * @param {any} body
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        callEncrypted: (activityId, body, options = {}) => __awaiter(this, void 0, void 0, function* () {
            // verify required parameter 'activityId' is not null or undefined
            if (activityId === null || activityId === undefined) {
                throw new base_1.RequiredError('activityId', 'Required parameter activityId was null or undefined when calling callEncrypted.');
            }
            // verify required parameter 'body' is not null or undefined
            if (body === null || body === undefined) {
                throw new base_1.RequiredError('body', 'Required parameter body was null or undefined when calling callEncrypted.');
            }
            const localVarPath = `/activity/{activityId}/encrypted`
                .replace(`{${"activityId"}}`, encodeURIComponent(String(activityId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, 'https://example.com');
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication app_key required
            // http bearer authentication required
            if (configuration && configuration.accessToken) {
                const accessToken = typeof configuration.accessToken === 'function'
                    ? yield configuration.accessToken()
                    : yield configuration.accessToken;
                localVarHeaderParameter["Authorization"] = "Bearer " + accessToken;
            }
            localVarHeaderParameter['Content-Type'] = 'application/octet-stream';
            const queryParameters = new URLSearchParams(localVarUrlObj.search);
            for (const key in localVarQueryParameter) {
                queryParameters.set(key, localVarQueryParameter[key]);
            }
            for (const key in options.query) {
                queryParameters.set(key, options.query[key]);
            }
            localVarUrlObj.search = (new URLSearchParams(queryParameters)).toString();
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            const nonString = typeof body !== 'string';
            const needsSerialization = nonString && configuration && configuration.isJsonMime
                ? configuration.isJsonMime(localVarRequestOptions.headers['Content-Type'])
                : nonString;
            localVarRequestOptions.data = needsSerialization
                ? JSON.stringify(body !== undefined ? body : {})
                : (body || "");
            return {
                url: localVarUrlObj.pathname + localVarUrlObj.search + localVarUrlObj.hash,
                options: localVarRequestOptions,
            };
        }),
        /**
         * **Note:** This call shall get routed as a provider event (see ProviderEvent structure).
         * @summary Creates new Activity based on given Agreement.
         * @param {string | CreateActivityRequest} stringCreateActivityRequest
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createActivity: (stringCreateActivityRequest, options = {}) => __awaiter(this, void 0, void 0, function* () {
            // verify required parameter 'stringCreateActivityRequest' is not null or undefined
            if (stringCreateActivityRequest === null || stringCreateActivityRequest === undefined) {
                throw new base_1.RequiredError('stringCreateActivityRequest', 'Required parameter stringCreateActivityRequest was null or undefined when calling createActivity.');
            }
            const localVarPath = `/activity`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, 'https://example.com');
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication app_key required
            // http bearer authentication required
            if (configuration && configuration.accessToken) {
                const accessToken = typeof configuration.accessToken === 'function'
                    ? yield configuration.accessToken()
                    : yield configuration.accessToken;
                localVarHeaderParameter["Authorization"] = "Bearer " + accessToken;
            }
            localVarHeaderParameter['Content-Type'] = 'application/json';
            const queryParameters = new URLSearchParams(localVarUrlObj.search);
            for (const key in localVarQueryParameter) {
                queryParameters.set(key, localVarQueryParameter[key]);
            }
            for (const key in options.query) {
                queryParameters.set(key, options.query[key]);
            }
            localVarUrlObj.search = (new URLSearchParams(queryParameters)).toString();
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            const nonString = typeof stringCreateActivityRequest !== 'string';
            const needsSerialization = nonString && configuration && configuration.isJsonMime
                ? configuration.isJsonMime(localVarRequestOptions.headers['Content-Type'])
                : nonString;
            localVarRequestOptions.data = needsSerialization
                ? JSON.stringify(stringCreateActivityRequest !== undefined ? stringCreateActivityRequest : {})
                : (stringCreateActivityRequest || "");
            return {
                url: localVarUrlObj.pathname + localVarUrlObj.search + localVarUrlObj.hash,
                options: localVarRequestOptions,
            };
        }),
        /**
         * **Note:** This call shall get routed as a provider event (see ProviderEvent structure).
         * @summary Destroys given Activity.
         * @param {string} activityId
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        destroyActivity: (activityId, options = {}) => __awaiter(this, void 0, void 0, function* () {
            // verify required parameter 'activityId' is not null or undefined
            if (activityId === null || activityId === undefined) {
                throw new base_1.RequiredError('activityId', 'Required parameter activityId was null or undefined when calling destroyActivity.');
            }
            const localVarPath = `/activity/{activityId}`
                .replace(`{${"activityId"}}`, encodeURIComponent(String(activityId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, 'https://example.com');
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'DELETE' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication app_key required
            // http bearer authentication required
            if (configuration && configuration.accessToken) {
                const accessToken = typeof configuration.accessToken === 'function'
                    ? yield configuration.accessToken()
                    : yield configuration.accessToken;
                localVarHeaderParameter["Authorization"] = "Bearer " + accessToken;
            }
            const queryParameters = new URLSearchParams(localVarUrlObj.search);
            for (const key in localVarQueryParameter) {
                queryParameters.set(key, localVarQueryParameter[key]);
            }
            for (const key in options.query) {
                queryParameters.set(key, options.query[key]);
            }
            localVarUrlObj.search = (new URLSearchParams(queryParameters)).toString();
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: localVarUrlObj.pathname + localVarUrlObj.search + localVarUrlObj.hash,
                options: localVarRequestOptions,
            };
        }),
        /**
         * **Note:** This call shall get routed directly to ExeUnit.
         * @summary Executes an ExeScript batch within a given Activity.
         * @param {string} activityId
         * @param {ExeScriptRequest} script
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        exec: (activityId, script, options = {}) => __awaiter(this, void 0, void 0, function* () {
            // verify required parameter 'activityId' is not null or undefined
            if (activityId === null || activityId === undefined) {
                throw new base_1.RequiredError('activityId', 'Required parameter activityId was null or undefined when calling exec.');
            }
            // verify required parameter 'script' is not null or undefined
            if (script === null || script === undefined) {
                throw new base_1.RequiredError('script', 'Required parameter script was null or undefined when calling exec.');
            }
            const localVarPath = `/activity/{activityId}/exec`
                .replace(`{${"activityId"}}`, encodeURIComponent(String(activityId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, 'https://example.com');
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication app_key required
            // http bearer authentication required
            if (configuration && configuration.accessToken) {
                const accessToken = typeof configuration.accessToken === 'function'
                    ? yield configuration.accessToken()
                    : yield configuration.accessToken;
                localVarHeaderParameter["Authorization"] = "Bearer " + accessToken;
            }
            localVarHeaderParameter['Content-Type'] = 'application/json';
            const queryParameters = new URLSearchParams(localVarUrlObj.search);
            for (const key in localVarQueryParameter) {
                queryParameters.set(key, localVarQueryParameter[key]);
            }
            for (const key in options.query) {
                queryParameters.set(key, options.query[key]);
            }
            localVarUrlObj.search = (new URLSearchParams(queryParameters)).toString();
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            const nonString = typeof script !== 'string';
            const needsSerialization = nonString && configuration && configuration.isJsonMime
                ? configuration.isJsonMime(localVarRequestOptions.headers['Content-Type'])
                : nonString;
            localVarRequestOptions.data = needsSerialization
                ? JSON.stringify(script !== undefined ? script : {})
                : (script || "");
            return {
                url: localVarUrlObj.pathname + localVarUrlObj.search + localVarUrlObj.hash,
                options: localVarRequestOptions,
            };
        }),
        /**
         * \'This call shall collect ExeScriptCommand result objects received directly from ExeUnit (via the long polling pattern). **Note:** two formats of response are specified (as indicated by the Accept header):   - application/json - standard JSON response, specified below, as code generators handle it properly.   - text/event-stream - an EventSource implementation (as per https://www.w3.org/TR/eventsource/).     This isn\'t explicitly specified as code generators generally are unable to handle this.     The streaming events adhere to following format:        event: runtime       data: <RuntimeEvent structure>      This streaming endpoint requires dedicated implementation. \'
         * @summary Queries for ExeScript batch results.
         * @param {string} activityId
         * @param {string} batchId
         * @param {number} [commandIndex] Wait until command with the specified index finishes. Must be accompanied by a valid \&quot;pollTimeout\&quot; query parameter.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getExecBatchResults: (activityId, batchId, commandIndex, options = {}) => __awaiter(this, void 0, void 0, function* () {
            // verify required parameter 'activityId' is not null or undefined
            if (activityId === null || activityId === undefined) {
                throw new base_1.RequiredError('activityId', 'Required parameter activityId was null or undefined when calling getExecBatchResults.');
            }
            // verify required parameter 'batchId' is not null or undefined
            if (batchId === null || batchId === undefined) {
                throw new base_1.RequiredError('batchId', 'Required parameter batchId was null or undefined when calling getExecBatchResults.');
            }
            const localVarPath = `/activity/{activityId}/exec/{batchId}`
                .replace(`{${"activityId"}}`, encodeURIComponent(String(activityId)))
                .replace(`{${"batchId"}}`, encodeURIComponent(String(batchId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, 'https://example.com');
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication app_key required
            // http bearer authentication required
            if (configuration && configuration.accessToken) {
                const accessToken = typeof configuration.accessToken === 'function'
                    ? yield configuration.accessToken()
                    : yield configuration.accessToken;
                localVarHeaderParameter["Authorization"] = "Bearer " + accessToken;
            }
            if (commandIndex !== undefined) {
                localVarQueryParameter['commandIndex'] = commandIndex;
            }
            const queryParameters = new URLSearchParams(localVarUrlObj.search);
            for (const key in localVarQueryParameter) {
                queryParameters.set(key, localVarQueryParameter[key]);
            }
            for (const key in options.query) {
                queryParameters.set(key, options.query[key]);
            }
            localVarUrlObj.search = (new URLSearchParams(queryParameters)).toString();
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: localVarUrlObj.pathname + localVarUrlObj.search + localVarUrlObj.hash,
                options: localVarRequestOptions,
            };
        }),
    };
};
/**
 * RequestorControlApi - functional programming interface
 * @export
 */
exports.RequestorControlApiFp = function (configuration) {
    return {
        /**
         *
         * @summary Sends encrypted command to secure exe-unit.
         * @param {string} activityId
         * @param {any} body
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        callEncrypted(activityId, body, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield exports.RequestorControlApiAxiosParamCreator(configuration).callEncrypted(activityId, body, options);
                return (axios = axios_1.default, basePath = base_1.BASE_PATH) => {
                    const axiosRequestArgs = Object.assign(Object.assign({}, localVarAxiosArgs.options), { url: basePath + localVarAxiosArgs.url });
                    return axios.request(axiosRequestArgs);
                };
            });
        },
        /**
         * **Note:** This call shall get routed as a provider event (see ProviderEvent structure).
         * @summary Creates new Activity based on given Agreement.
         * @param {string | CreateActivityRequest} stringCreateActivityRequest
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createActivity(stringCreateActivityRequest, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield exports.RequestorControlApiAxiosParamCreator(configuration).createActivity(stringCreateActivityRequest, options);
                return (axios = axios_1.default, basePath = base_1.BASE_PATH) => {
                    const axiosRequestArgs = Object.assign(Object.assign({}, localVarAxiosArgs.options), { url: basePath + localVarAxiosArgs.url });
                    return axios.request(axiosRequestArgs);
                };
            });
        },
        /**
         * **Note:** This call shall get routed as a provider event (see ProviderEvent structure).
         * @summary Destroys given Activity.
         * @param {string} activityId
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        destroyActivity(activityId, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield exports.RequestorControlApiAxiosParamCreator(configuration).destroyActivity(activityId, options);
                return (axios = axios_1.default, basePath = base_1.BASE_PATH) => {
                    const axiosRequestArgs = Object.assign(Object.assign({}, localVarAxiosArgs.options), { url: basePath + localVarAxiosArgs.url });
                    return axios.request(axiosRequestArgs);
                };
            });
        },
        /**
         * **Note:** This call shall get routed directly to ExeUnit.
         * @summary Executes an ExeScript batch within a given Activity.
         * @param {string} activityId
         * @param {ExeScriptRequest} script
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        exec(activityId, script, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield exports.RequestorControlApiAxiosParamCreator(configuration).exec(activityId, script, options);
                return (axios = axios_1.default, basePath = base_1.BASE_PATH) => {
                    const axiosRequestArgs = Object.assign(Object.assign({}, localVarAxiosArgs.options), { url: basePath + localVarAxiosArgs.url });
                    return axios.request(axiosRequestArgs);
                };
            });
        },
        /**
         * \'This call shall collect ExeScriptCommand result objects received directly from ExeUnit (via the long polling pattern). **Note:** two formats of response are specified (as indicated by the Accept header):   - application/json - standard JSON response, specified below, as code generators handle it properly.   - text/event-stream - an EventSource implementation (as per https://www.w3.org/TR/eventsource/).     This isn\'t explicitly specified as code generators generally are unable to handle this.     The streaming events adhere to following format:        event: runtime       data: <RuntimeEvent structure>      This streaming endpoint requires dedicated implementation. \'
         * @summary Queries for ExeScript batch results.
         * @param {string} activityId
         * @param {string} batchId
         * @param {number} [commandIndex] Wait until command with the specified index finishes. Must be accompanied by a valid \&quot;pollTimeout\&quot; query parameter.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getExecBatchResults(activityId, batchId, commandIndex, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield exports.RequestorControlApiAxiosParamCreator(configuration).getExecBatchResults(activityId, batchId, commandIndex, options);
                return (axios = axios_1.default, basePath = base_1.BASE_PATH) => {
                    const axiosRequestArgs = Object.assign(Object.assign({}, localVarAxiosArgs.options), { url: basePath + localVarAxiosArgs.url });
                    return axios.request(axiosRequestArgs);
                };
            });
        },
    };
};
/**
 * RequestorControlApi - factory interface
 * @export
 */
exports.RequestorControlApiFactory = function (configuration, basePath, axios) {
    return {
        /**
         *
         * @summary Sends encrypted command to secure exe-unit.
         * @param {string} activityId
         * @param {any} body
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        callEncrypted(activityId, body, options) {
            return exports.RequestorControlApiFp(configuration).callEncrypted(activityId, body, options).then((request) => request(axios, basePath));
        },
        /**
         * **Note:** This call shall get routed as a provider event (see ProviderEvent structure).
         * @summary Creates new Activity based on given Agreement.
         * @param {string | CreateActivityRequest} stringCreateActivityRequest
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createActivity(stringCreateActivityRequest, options) {
            return exports.RequestorControlApiFp(configuration).createActivity(stringCreateActivityRequest, options).then((request) => request(axios, basePath));
        },
        /**
         * **Note:** This call shall get routed as a provider event (see ProviderEvent structure).
         * @summary Destroys given Activity.
         * @param {string} activityId
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        destroyActivity(activityId, options) {
            return exports.RequestorControlApiFp(configuration).destroyActivity(activityId, options).then((request) => request(axios, basePath));
        },
        /**
         * **Note:** This call shall get routed directly to ExeUnit.
         * @summary Executes an ExeScript batch within a given Activity.
         * @param {string} activityId
         * @param {ExeScriptRequest} script
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        exec(activityId, script, options) {
            return exports.RequestorControlApiFp(configuration).exec(activityId, script, options).then((request) => request(axios, basePath));
        },
        /**
         * \'This call shall collect ExeScriptCommand result objects received directly from ExeUnit (via the long polling pattern). **Note:** two formats of response are specified (as indicated by the Accept header):   - application/json - standard JSON response, specified below, as code generators handle it properly.   - text/event-stream - an EventSource implementation (as per https://www.w3.org/TR/eventsource/).     This isn\'t explicitly specified as code generators generally are unable to handle this.     The streaming events adhere to following format:        event: runtime       data: <RuntimeEvent structure>      This streaming endpoint requires dedicated implementation. \'
         * @summary Queries for ExeScript batch results.
         * @param {string} activityId
         * @param {string} batchId
         * @param {number} [commandIndex] Wait until command with the specified index finishes. Must be accompanied by a valid \&quot;pollTimeout\&quot; query parameter.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getExecBatchResults(activityId, batchId, commandIndex, options) {
            return exports.RequestorControlApiFp(configuration).getExecBatchResults(activityId, batchId, commandIndex, options).then((request) => request(axios, basePath));
        },
    };
};
/**
 * RequestorControlApi - object-oriented interface
 * @export
 * @class RequestorControlApi
 * @extends {BaseAPI}
 */
class RequestorControlApi extends base_1.BaseAPI {
    /**
     *
     * @summary Sends encrypted command to secure exe-unit.
     * @param {string} activityId
     * @param {any} body
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof RequestorControlApi
     */
    callEncrypted(activityId, body, options) {
        return exports.RequestorControlApiFp(this.configuration).callEncrypted(activityId, body, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * **Note:** This call shall get routed as a provider event (see ProviderEvent structure).
     * @summary Creates new Activity based on given Agreement.
     * @param {string | CreateActivityRequest} stringCreateActivityRequest
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof RequestorControlApi
     */
    createActivity(stringCreateActivityRequest, options) {
        return exports.RequestorControlApiFp(this.configuration).createActivity(stringCreateActivityRequest, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * **Note:** This call shall get routed as a provider event (see ProviderEvent structure).
     * @summary Destroys given Activity.
     * @param {string} activityId
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof RequestorControlApi
     */
    destroyActivity(activityId, options) {
        return exports.RequestorControlApiFp(this.configuration).destroyActivity(activityId, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * **Note:** This call shall get routed directly to ExeUnit.
     * @summary Executes an ExeScript batch within a given Activity.
     * @param {string} activityId
     * @param {ExeScriptRequest} script
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof RequestorControlApi
     */
    exec(activityId, script, options) {
        return exports.RequestorControlApiFp(this.configuration).exec(activityId, script, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * \'This call shall collect ExeScriptCommand result objects received directly from ExeUnit (via the long polling pattern). **Note:** two formats of response are specified (as indicated by the Accept header):   - application/json - standard JSON response, specified below, as code generators handle it properly.   - text/event-stream - an EventSource implementation (as per https://www.w3.org/TR/eventsource/).     This isn\'t explicitly specified as code generators generally are unable to handle this.     The streaming events adhere to following format:        event: runtime       data: <RuntimeEvent structure>      This streaming endpoint requires dedicated implementation. \'
     * @summary Queries for ExeScript batch results.
     * @param {string} activityId
     * @param {string} batchId
     * @param {number} [commandIndex] Wait until command with the specified index finishes. Must be accompanied by a valid \&quot;pollTimeout\&quot; query parameter.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof RequestorControlApi
     */
    getExecBatchResults(activityId, batchId, commandIndex, options) {
        return exports.RequestorControlApiFp(this.configuration).getExecBatchResults(activityId, batchId, commandIndex, options).then((request) => request(this.axios, this.basePath));
    }
}
exports.RequestorControlApi = RequestorControlApi;
