// yajsapi/dist/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sleep = exports.range = exports.Queue = exports.promisify = exports.logUtils = exports.logger = exports.getAllProperties = exports.changeLogLevel = exports.eventLoop = exports.CancellationToken = exports.AsyncWrapper = exports.asyncWith = exports.AsyncExitStack = exports.applyMixins = void 0;
const applyMixins_1 = __importDefault(require("./applyMixins"));
exports.applyMixins = applyMixins_1.default;
const asyncExitStack_1 = __importDefault(require("./asyncExitStack"));
exports.AsyncExitStack = asyncExitStack_1.default;
const asyncWith_1 = __importDefault(require("./asyncWith"));
exports.asyncWith = asyncWith_1.default;
const asyncWrapper_1 = __importDefault(require("./asyncWrapper"));
exports.AsyncWrapper = asyncWrapper_1.default;
const cancellationToken_1 = __importDefault(require("./cancellationToken"));
exports.CancellationToken = cancellationToken_1.default;
const eventLoop_1 = __importDefault(require("./eventLoop"));
exports.eventLoop = eventLoop_1.default;
const getAllProperties_1 = __importDefault(require("./getAllProperties"));
exports.getAllProperties = getAllProperties_1.default;
const log_1 = __importStar(require("./log")), logUtils = log_1;
exports.logger = log_1.default;
exports.logUtils = logUtils;
const log_2 = require("./log");
Object.defineProperty(exports, "changeLogLevel", { enumerable: true, get: function () { return log_2.changeLogLevel; } });
const promisify_1 = __importDefault(require("./promisify"));
exports.promisify = promisify_1.default;
const queue_1 = __importDefault(require("./queue"));
exports.Queue = queue_1.default;
const range_1 = __importDefault(require("./range"));
exports.range = range_1.default;
const sleep_1 = __importDefault(require("./sleep"));
exports.sleep = sleep_1.default;
