// yajsapi/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = exports.storage = exports.rest = exports.props = exports.WorkContext = exports.vm = exports.sgx = exports.Task = exports.Executor = void 0;
const executor_1 = require("./executor");
Object.defineProperty(exports, "Executor", { enumerable: true, get: function () { return executor_1.Executor; } });
Object.defineProperty(exports, "Task", { enumerable: true, get: function () { return executor_1.Task; } });
Object.defineProperty(exports, "sgx", { enumerable: true, get: function () { return executor_1.sgx; } });
Object.defineProperty(exports, "vm", { enumerable: true, get: function () { return executor_1.vm; } });
const ctx_1 = require("./executor/ctx");
Object.defineProperty(exports, "WorkContext", { enumerable: true, get: function () { return ctx_1.WorkContext; } });
const props = __importStar(require("./props"));
exports.props = props;
const rest = __importStar(require("./rest"));
exports.rest = rest;
const storage = __importStar(require("./storage"));
exports.storage = storage;
const utils = __importStar(require("./utils"));
exports.utils = utils;
