// yajsapi/dist/rest/payment.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payment = exports.Allocation = exports.DebitNote = exports.InvoiceStatus = exports.Invoice = void 0;
const dayjs_1 = __importDefault(require("dayjs"));
const utc_1 = __importDefault(require("dayjs/plugin/utc"));
const resource_1 = require("./resource");
const yap = __importStar(require("ya-ts-client/dist/ya-payment/src/models"));
const api_1 = require("ya-ts-client/dist/ya-payment/api");
const utils_1 = require("../utils");
dayjs_1.default.extend(utc_1.default);
class yInvoice {
}
class yDebitNote {
}
class Invoice extends yInvoice {
    constructor(_api, _base) {
        super();
        for (let [key, value] of Object.entries(_base)) {
            this[key] = value;
        }
        this._api = _api;
    }
    async accept(amount, allocation) {
        let acceptance = {
            totalAmountAccepted: amount.toString(),
            allocationId: allocation.id,
        };
        acceptance.totalAmountAccepted = amount.toString();
        acceptance.allocationId = allocation.id;
        await this._api.acceptInvoice(this.invoiceId, acceptance, undefined, { timeout: 15000 });
    }
}
exports.Invoice = Invoice;
exports.InvoiceStatus = yap.InvoiceStatus;
class DebitNote extends yDebitNote {
    constructor(_api, _base) {
        super();
        for (let [key, value] of Object.entries(_base)) {
            this[key] = value;
        }
        this._api = _api;
    }
    async accept(amount, allocation) {
        let acceptance = {
            totalAmountAccepted: amount.toString(),
            allocationId: allocation.id,
        };
        acceptance.totalAmountAccepted = amount.toString();
        acceptance.allocationId = allocation.id;
        await this._api.acceptDebitNote(this.debitNoteId, acceptance, undefined, { timeout: 15000 });
    }
}
exports.DebitNote = DebitNote;
class _Link {
}
class AllocationDetails {
}
class Allocation extends _Link {
    async details() {
        let allocationDetails = new AllocationDetails();
        try {
            let { data: details, } = await this._api.getAllocation(this.id, { timeout: 15000 });
            allocationDetails.spent_amount = parseFloat(details.spentAmount);
            allocationDetails.remaining_amount = parseFloat(details.remainingAmount);
        }
        catch (error) {
            utils_1.logger.error(error);
            throw new Error(error);
        }
        return allocationDetails;
    }
    async delete() {
        try {
            await this._api.releaseAllocation(this.id, { timeout: 15000 });
        }
        catch (error) {
            utils_1.logger.error(`Release allocation: ${error}`);
        }
    }
}
exports.Allocation = Allocation;
class _AllocationTask extends resource_1.ResourceCtx {
    constructor(api, model) {
        super();
        this._api = api;
        this.model = model;
    }
    async ready() {
        try {
            let { data: new_allocation, } = await this._api.createAllocation(this.model, undefined, undefined, { timeout: 25000 });
            this._id = new_allocation.allocationId;
            let model = this.model;
            if (!model.totalAmount)
                throw "";
            if (!model.timeout)
                throw "";
            if (!this._id)
                throw "";
            let _allocation = new Allocation();
            _allocation.id = this._id;
            _allocation._api = this._api;
            _allocation.payment_platform = model.paymentPlatform;
            _allocation.payment_address = model.address;
            _allocation.amount = parseFloat(model.totalAmount);
            _allocation.expires = new Date(parseInt(model.timeout) * 1000);
            return _allocation;
        }
        catch (error) {
            const msg = error.response && error.response.data ? error.response.data.message : error.message;
            utils_1.logger.error(`Payment allocation error (message: ${msg}). Please run "yagna payment status" to check your account.`);
            throw new Error(error);
        }
    }
    async done() {
        if (this._id) {
            try {
                await this._api.releaseAllocation(this._id, { timeout: 5000 });
            }
            catch (error) {
                utils_1.logger.error(`Release allocation: ${error}`);
            }
        }
    }
}
class yAllocation {
    constructor() {
        this.allocationId = "";
    }
}
class Payment {
    constructor(cfg) {
        this._api = new api_1.RequestorApi(cfg);
    }
    new_allocation(amount, payment_platform, payment_address, expires = null, make_deposit = false) {
        let allocation_timeout = expires || dayjs_1.default().add(30, "minute").utc().toDate();
        let _allocation = new yAllocation();
        _allocation.paymentPlatform = payment_platform;
        _allocation.address = payment_address;
        _allocation.totalAmount = amount.toString();
        _allocation.timeout = allocation_timeout.toISOString();
        _allocation.makeDeposit = make_deposit;
        _allocation.spentAmount = "";
        _allocation.remainingAmount = "";
        return new _AllocationTask(this._api, _allocation);
    }
    async *allocations() {
        try {
            let { data: result } = await this._api.getAllocations();
            for (let alloc_obj of result) {
                let _allocation = new Allocation();
                _allocation._api = this._api;
                _allocation.id = alloc_obj.allocationId;
                _allocation.amount = parseFloat(alloc_obj.totalAmount);
                _allocation.payment_platform = alloc_obj.paymentPlatform;
                _allocation.payment_address = alloc_obj.address;
                _allocation.expires = new Date(parseInt(alloc_obj.timeout) * 1000);
                yield _allocation;
            }
        }
        catch (error) {
            throw error;
        }
        return;
    }
    async allocation(allocation_id) {
        try {
            let { data: result, } = await this._api.getAllocation(allocation_id);
            let allocation_obj = result;
            let _allocation = new Allocation();
            _allocation._api = this._api;
            _allocation.id = allocation_obj.allocationId;
            _allocation.amount = parseFloat(allocation_obj.totalAmount);
            _allocation.payment_platform = allocation_obj.paymentPlatform;
            _allocation.payment_address = allocation_obj.address;
            _allocation.expires = new Date(parseInt(allocation_obj.timeout) * 1000);
            return _allocation;
        }
        catch (error) {
            throw error;
        }
    }
    async *accounts() {
        let { data: _accounts } = await this._api.getRequestorAccounts();
        for (let account_obj of _accounts) {
            yield account_obj;
        }
    }
    async decorate_demand(ids) {
        const { data: _decorate_demand } = await this._api.getDemandDecorations(ids);
        return _decorate_demand;
    }
    async debit_note(debit_note_id) {
        let { data: debit_note_obj } = await this._api.getDebitNote(debit_note_id, { timeout: 5000 });
        return new DebitNote(this._api, debit_note_obj);
    }
    async *invoices() {
        let { data: result } = await this._api.getInvoices(undefined, undefined, { timeout: 5000 });
        for (let invoice_obj of result) {
            yield new Invoice(this._api, invoice_obj);
        }
        return;
    }
    async invoice(invoice_id) {
        let { data: invoice_obj } = await this._api.getInvoice(invoice_id, { timeout: 5000 });
        return new Invoice(this._api, invoice_obj);
    }
    incoming_invoices(cancellationToken) {
        let ts = dayjs_1.default().utc();
        let api = this._api;
        let self = this;
        async function* fetch(init_ts) {
            let ts = init_ts;
            while (true) {
                if (cancellationToken.cancelled)
                    break;
                try {
                    let { data: events } = await api.getInvoiceEvents(5, ts.format("YYYY-MM-DDTHH:mm:ss.SSSSSSZ"));
                    for (let ev of events) {
                        utils_1.logger.debug(`Received invoice event: ${JSON.stringify(ev)}, type: ${JSON.stringify(Object.getPrototypeOf(ev))}`);
                        ts = dayjs_1.default(ev.eventDate);
                        if (ev.eventType === "InvoiceReceivedEvent") {
                            let invoice = await self.invoice(ev["invoiceId"]);
                            yield invoice;
                        }
                    }
                    if (!events || !events.length) {
                        await utils_1.sleep(1);
                    }
                }
                catch (error) {
                    utils_1.logger.error(`Received invoice error: ${error}`);
                }
            }
            return;
        }
        return fetch(ts);
    }
    incoming_debit_notes(cancellationToken) {
        let ts = dayjs_1.default().utc();
        let api = this._api;
        let self = this;
        async function* fetch(init_ts) {
            let ts = init_ts;
            while (true) {
                if (cancellationToken.cancelled)
                    break;
                try {
                    let { data: events } = await api.getDebitNoteEvents(5, ts.format("YYYY-MM-DDTHH:mm:ss.SSSSSSZ"));
                    for (let ev of events) {
                        utils_1.logger.debug(`Received debit note event: ${JSON.stringify(ev)}, type: ${JSON.stringify(Object.getPrototypeOf(ev))}`);
                        ts = dayjs_1.default(ev.eventDate);
                        if (ev.eventType === "DebitNoteReceivedEvent") {
                            let debit_note = await self.debit_note(ev["debitNoteId"]);
                            yield debit_note;
                        }
                    }
                    if (!events || !events.length) {
                        await utils_1.sleep(1);
                    }
                }
                catch (error) {
                    utils_1.logger.error(`Received debit note error: ${error}`);
                }
            }
            return;
        }
        return fetch(ts);
    }
}
exports.Payment = Payment;
